[클린코드 2기 임영서] 계산기 미션

youngseo-im opened the pull request:
안녕하세요 : ) 코드리뷰 요청드립니다. 

또 추가로 궁금한점이 있어서 질문남깁니다.
https://www.youtube.com/watch?v=fzrT3eoecUw 
위 동영상에서 최신 코드를 받아올때 fetch나 pull이 아닌 rebase를 이용하더라구요
평소에 rebase를 사용해보지않았어서 검색해서 찾아보다가 아래와 같이 정리를했습니다.

아래와 같은 이유로 현재 코드리뷰 방식에는 rebase를 통해서 최신코드를 받아오는 방법이 적합하다는건 알았지만 어떤 경우에 pull이나 fetch 명령어를 사용하는게 적합한지에 대해서는 의문이 좀 생겼습니다.

 그냥 단순하게 브렌치의 개수가 많지않다면 pull이나 fetch 명령어를 사용해서 병합한다고 이해하면될까요 ?? 
----------------------------------------------------------------------------

1. git fetch => 최신 코드를 받아오기만 하기때문에 추가로 merge하는 작업이 필요함
2. git pull의 경우 git fetch + merge가 합쳐진 동작

- 1,2번 케이스의 경우 merge되는 방식은 branch의 최종 결과만을 가지고 합병

3. rebase를 활용할 경우 지정한 브렌치 즉 마스터 브렌치에서 분기된 내이름으로 생성된 브렌치를 베이스로 기준삼아서 합병할 수 있다
   1. 중복 수정된 로그가 남지않기때문에 history가 단순해짐
   2. 브랜치의 변경사항을 순서대로 다른 브랜치에 적용하면서 합병함

youngseo-im committed:
[feat] 계산기 사칙연산 로직 구현

JunilHwang commented on src/js/index.js:
```diff
@@ -0,0 +1,83 @@
+const operatorBtn = document.getElementById("operator");
+
+const totalCountToArr = (unit) => totalCount.innerHTML.split(unit);
+const isOperator = () => isNaN(Number(value));
```
일단 value가 현재 정의 되지 않은 상태라서, 앱을 실행하고 계산기를 클릭하면 오류가 발생하고 있습니다.
![image](https://user-images.githubusercontent.com/18749057/157446080-ac2b3a7e-fd56-417f-93ab-4090f45e90d6.png)

youngseo-im replied:
함수를 분리하면서 놓친 것 같습니다 이번 코드 리뷰 때 수정했습니다 .. ㅎㅎ 앞으로는 좀 더 꼼꼼히 테스트 해봐야겠네요 ㅠㅠ

JunilHwang commented on src/js/index.js:
```diff
@@ -0,0 +1,83 @@
+const totalCountToArr = (unit) => totalCount.innerHTML.split(unit);
+const isOperator = () => isNaN(Number(value));
+
+const isMaxLength = () => {
+  const curOperator = totalCountToArr("").find((v) => isOperator(v));
+  if (curOperator) {
+    return totalCountToArr(curOperator)[1].split("").length >= 3 ? true : false;
+  } else {
+    return totalCountToArr("").length >= 3 ? true : false;
+  }
+};
```
```suggestion
const isMaxLength = () => {
  const curOperator = totalCountToArr("").find((v) => isOperator(v));
  if (curOperator) {
    return totalCountToArr(curOperator)[1].split("").length >= 3;
  }
  return totalCountToArr("").length >= 3;
};
```
이 부분은 이렇게 표현할 수 있겠네요! 그리고 if로 진입시 바로 return 하기 때문에 else 구문은 불필요해보입니다.

JunilHwang replied:
이 코드의 경우 영서님이 아닌 다른 사람이 봤을 때 이게 정확히 무엇에 대한 isMaxLength인지 가늠하기 힘들지 않을까요?

- 이 코드는 실제로 totalCount라는 dom의 길이가 maxLength인지 판단하고 있습니다.
- 그리고 maxLength는 3인데 함수 내부에서만 표현되고 있습니다.
- 즉, 이 함수의 이름만 봤을 때 무슨 용도인지 파악하기란 쉽지 않은 상태인거죠

이 과정은 `클린코드` 과정입니다.
- 내가 작성한 코드가 유지보수하기 쉬운가
- 유지보수 하기 쉽다는 것은
- 코드를 읽기 쉽다는 것
- 코드를 이해하기 쉽다는 것
- 내가 아닌 다른 사람도 이 코드를 보고 쉽게 수정할 수 있다는 것

그렇기 때문에 위에 대한 고민을 지속적으로 해보시는게 좋습니다 ㅎㅎ

youngseo-im replied:
준일님 피드백을 받고서 많은 생각이 들었습니다. 
내가 아닌 다른사람이 봤을 때 무엇을 하는 함수인지 알기 쉬워야한다.
항상 고민을 하지만 다시볼때마다 아쉽기도하고 아직도 많이 어려운 부분인 것 같습니다.
피드백을 토대로 코드를 수정했지만 이 부분이 말끔하게 해결된건지는 아직 의문입니다 ㅠㅠ

JunilHwang commented on src/js/index.js:
```diff
@@ -0,0 +1,83 @@
+const isOperator = () => isNaN(Number(value));
+
+const isMaxLength = () => {
+  const curOperator = totalCountToArr("").find((v) => isOperator(v));
```
```suggestion
  const curOperator = totalCountToArr("").find(isOperator);
```
이렇게 표현할 수 있답니다!

JunilHwang replied:
그리고 `totalCountToArr("")` 의 경우 변수로 담아서 사용할 수 있지 않을까요?

youngseo-im replied:
피드백대로 변수로 담아서 사용되도록 수정했습니다 : )

JunilHwang commented on src/js/index.js:
```diff
@@ -0,0 +1,83 @@
+};
+
+const addText = (text) => {
+  const isFirstClick = totalCountToArr("")[0] === "0" ? true : false;
```
```suggestion
  const isFirstClick = totalCountToArr("")[0] === "0";
```
이렇게 표현할 수 있습니다.

JunilHwang commented on src/js/index.js:
```diff
@@ -0,0 +1,83 @@
+};
+
+const onErrHandler = (type) => {
+  switch (type) {
+    case "duplication operator":
+      alert("숫자를 먼저 입력한 후 연산자를 입력해주세요!");
+      break;
+    case "max Length":
+      alert("숫자는 세 자리까지만 입력 가능합니다!");
+      break;
+    default:
+      return;
```
`duplication operator`와 `max Length`는 상수로 선언하여 사용하면 어떨까요? 그래야 코드를 작성할 때 실수를 예방할 수 있답니다!

youngseo-im replied:
일하면서 오타로 인해 시간 낭비를 한 적이 몇 번 있었는데 
이걸 상수로 관리해야한다는 생각을 왜 못했을까요 ..
새겨듣겠습니다 ㅎㅎ

JunilHwang commented on src/js/index.js:
```diff
@@ -0,0 +1,83 @@
+const isMaxLength = () => {
+  const curOperator = totalCountToArr("").find((v) => isOperator(v));
+  if (curOperator) {
+    return totalCountToArr(curOperator)[1].split("").length >= 3 ? true : false;
+  } else {
+    return totalCountToArr("").length >= 3 ? true : false;
```
3의 경우 상수로 선언해서 사용하면 어떨까요?
가령, maxLength가 5나 10으로 변경된다면 3이라는 키워드를 찾아서 다 수정해야 하는데,
3을 `MAX_LENGTH`라는 변수로 선언한다면, `MAX_LENGTH`의 값만 수정하면 된답니다!

youngseo-im replied:
전부 대문자를 쓰는 컨벤션에대해서 이번기회에 검색을 해봤는데
상수와 전역변수의 경우 모두 대문자로 변수를 선언하는것이 권장하는 방식인가요 ??

JunilHwang replied:
보통 상수의 경우에만 전역변수로 사용한답니다!

JunilHwang commented on src/js/index.js:
```diff
@@ -0,0 +1,83 @@
+};
+
+const sum = () => {
+  const curOperator = totalCountToArr("").find((v) => isOperator(v));
+  const valueArr = totalCountToArr(curOperator).map((v) => Number(v));
```
```suggestion
  const curOperator = totalCountToArr("").find(isOperator);
  const valueArr = totalCountToArr(curOperator).map(Number);
```
이렇게 표현할 수 있답니다!

youngseo-im replied:
이렇게 사용할 수 있는지 처음 알았습니다 ㅎㅎ 또 하나 배워가네요 !

JunilHwang commented on src/js/index.js:
```diff
@@ -0,0 +1,83 @@
+  const curOperator = totalCountToArr("").find((v) => isOperator(v));
+  const valueArr = totalCountToArr(curOperator).map((v) => Number(v));
+
+  switch (curOperator) {
+    case "+":
+      return valueArr[0] + valueArr[1];
+    case "-":
+      return valueArr[0] - valueArr[1];
+    case "/":
+      return valueArr[0] / valueArr[1];
+    case "X":
+      return valueArr[0] * valueArr[1];
+  }
```
+, -, /, X 등의 키워드도 상수로 선언해서 관리해주면 어떨까요?

youngseo-im replied:
연산자의 키워드도 상수로 선언해서 관리했고 
swtich문은 lookup table이 가독성이 더 좋은 것 같아 수정했습니다 : )

JunilHwang reviewed (changes requested):
안녕하세요 영서님! 계산기 미션을 함께하게 된 리뷰어 황준일입니다 🙇‍♂️
먼저 이 과정을 수행하면서 항상 고민해야 하는 것은 `클린코드란 무엇인가` 입니다.
그리고 내가 작성한 코드가 `클린한가?` 에 대해서 계속 고민하고 코드를 작성해야합니다.

`계산기` 라는게 어떻게 보면 단순하기 때문에 쉽게 구현할 수 있습니다.
그런데 우리가 다양한 미션을 통해서 얻어가고자 하는 것이 `구현` 인걸까요?

어떤 방식으로 접근하냐에 따라서 챌린징의 수준이 달라지리라 생각합니다 ㅎㅎ

일단 `클린코드` 라는 것은
- 유지보수하기가 쉬운가?
- 유지보수하기 쉽다는 의미는 무엇인가?
- 다른 사람도 내 코드를 봤을 때 직관적으로 이해할 수 있는가?
- 특정 코드를 수정할 때 다른 코드에 여파가 심한가?
- 다른 코드에 여파가 심하다면, 그 여파를 줄일 수 있는가? (사이드이펙트)
- 여파가 심한 이유는 무엇인가?
- 내가 작성한 이름이 직관적인가?
- 최소한의 수정으로 코드를 유지할 수 있는가?

대충 이런 고민들이 있겠네요 ㅎㅎ
핵심 키워드는 `유지보수하기가 쉬운가` 입니다. 
여기에서 모든게 시작됩니다. 

다시 한 번 고민해봅시다!

미리 예고해드리자면, 저는 영서님을 쉽게 놓아드리지 않을 예정입니다 😁

---

그리고 git과 관련된 질문에 대해 답변드리자면,
merge의 경우 코드를 병합하는 것이고,
rebase의 경우 git log를 아예 합친다? 라고 볼 수 있습니다.

그래서 코드 관점으로 보면 사실 결과적으로는 똑같지만, rebase는 log를 합치기 때문에 대부분의 상황에서는 rebase가 더 좋습니다.

일단 이 과정을 수행하는 중에는 어떤 방식으로 하든 사실 크게 상관 없습니다 ㅎㅎ 
다만 현업에서는 rebase가 대부분의 경우를 제외하곤(ex: 하나의 브랜치에 동시에 작업하는 경우..)에 더 좋긴합니다!

youngseo-im committed:
[etc] gitignore DS_store 추가

youngseo-im committed:
[feat] 스크립트 태그에 type="module" 어트리뷰트 추가

youngseo-im committed:
[feat] define 파일 생성

youngseo-im committed:
[feat] 피드백 반영 코드 수정

youngseo-im committed:
[feat] 노드가 추가되기 전에 isErr의 값이 판단되지않도록 위치 수정

youngseo-im committed:
[build] cypress install

youngseo-im committed:
[feat] '=', '0' 상수화

youngseo-im committed:
[feat]
- isNaN => Number.isNaN 수정
- 첫 번째 노드가 0인지 판단하는 isFirstNodeIsZero 추가
- 양수, 음수 상관 없이 소숫점을 버릴 수 있도록 roundDown 함수 구현

youngseo-im committed:
[feat] 초기화 버튼 상수 적용

youngseo-im committed:
[feat] 성능과 보안에 취약한 innerHTML => textContent로 수정

youngseo-im committed:
[reafctor] switch => lookuptable 수정

youngseo-im committed:
[feat] cypress test를 위해 data-set attribute 추가

youngseo-im committed:
[feat] 7가지 테스트 케이스 구현
- 2개의 숫자에 대해 덧셈이 가능하다.
- 2개의 숫자에 대해 뺄셈이 가능하다.
- 2개의 숫자에 대해 곱셈이 가능하다.
- 2개의 숫자에 대해 나눗셈이 가능하다.
- AC 버튼을 누르면 0으로 초기화 된다.
- 숫자는 한 번에 최대 3자리 수 까지 입력이 가능하다.
- 소숫점 이하는 버림한다.

youngseo-im committed:
[refactor]
- 사칙연산 시 반복되는 로직 분리

youngseo-im commented:
늦은 시간에 리뷰 요청드려서 죄송합니다 ㅠㅠ
현재 와이파이가 없고 핫스팟 데이터를 다 사용해서 가족이 오기 전에는 리뷰 요청을 못할 것 같아
부득이하게 이 시간에 리뷰 요청을 해놨습니다 양해 부탁드려요 

코로나 확진으로 생각보다 몸이 좋지않아서 피드백 수정이 늦어서 많이 아쉽네요ㅠㅠ 

먼저 rebase에 대해서 답변해주셔서 감사합니다.

피드백을 반영해서 아래의 작성한 부분을 중점으로 리팩토링을 진행했습니다.
- 유지보수하기가 쉬운가를 중점으로 코드를 재작성해봤습니다 .
- 작성한 이름이 직관적일 수 있도록 함수는 최대한 한 가지 일만 할 수 있도록 다시 분리했습니다.

감사합니다 : )

JunilHwang commented on src/js/utils/define.js:
```diff
@@ -0,0 +1,8 @@
+export const duplicationOperator = "duplication operator";
+export const maxLength = "max length";
+export const plus = "+";
+export const minus = "-";
+export const division = "/";
+export const multiplication = "X";
+export const equal = "=";
+export const zero = "0";
```
보통 상수는 대문자로 정의한답니다!

JunilHwang commented on src/js/utils/define.js:
```diff
@@ -0,0 +1,8 @@
+export const duplicationOperator = "duplication operator";
+export const maxLength = "max length";
+export const plus = "+";
+export const minus = "-";
+export const division = "/";
+export const multiplication = "X";
+export const equal = "=";
```
```suggestion
export const Operator = {
  PLUS: "+",
  MINUS: "-",
  MULTIPLE: "X",
  DIVISION: "/",
  EQUALS: "=",
}
```
보통 이렇게 묶어서 상수로 표현한답니다!

JunilHwang commented on src/js/index.js:
```diff
@@ -0,0 +1,104 @@
+};
+
+const addNumber = (cur) => {
+  if (isFirstNodeIsZero()) {
+    totalCount.textContent = "";
+  }
+
+  totalCount.appendChild(document.createTextNode(cur));
```
```suggestion
  totalCount.innerHTML = Number(cur);
```
이렇게 표현하면 더 직관적이랍니다!

JunilHwang commented on src/js/index.js:
```diff
@@ -0,0 +1,104 @@
+  return Number(curCount) < MAX_VALUE;
+};
+
+const addNumber = (cur) => {
```
cur 보단 current 처럼 표현하는게 더 좋을 것 같아요! 줄여쓰게되면 정확한 의미를 유추하기 힘들때가 있습니다.

JunilHwang commented on src/js/index.js:
```diff
@@ -0,0 +1,104 @@
+  return firstNode === zero;
+};
+
+const isLessThanMaxValue = () => {
+  const MAX_VALUE = 1000;
+  const curCount = totalCount.textContent;
+  if (curOperator()) {
+    const [_, second] = valueList(curOperator());
+    return Number(second) < MAX_VALUE;
+  }
+  return Number(curCount) < MAX_VALUE;
+};
```
이 함수의 경우 도메인 로직과 뷰 로직이 뒤섞여 있습니다.
명확하게 분리해서 관리해주면 어떨까요?

쉽게 말해서 뷰에 대한 수행은 특정 함수에게 몰아주는게 제일 좋다고 생각합니다.

JunilHwang commented on src/js/index.js:
```diff
@@ -0,0 +1,104 @@
+const isOperator = (value) => isNaNCustom(value);
+const curOperator = () => totalCountToArr().find(isOperator);
+
+const isFirstNodeIsZero = () => {
+  const [firstNode] = totalCountToArr();
+  return firstNode === zero;
+};
```
이 함수 또한 앞선 리뷰와 마찬가지입니다.

JunilHwang commented on src/js/index.js:
```diff
@@ -0,0 +1,104 @@
+  }
+};
+
+const roundDown = (value) =>
+  value >= 0 ? Math.floor(value) : Math.ceil(value);
```
약간의 편법(?)으로 다음과 같이 표현할 수 있답니다.
```suggestion
const roundDown = (value) => ~~value;
```

`~` 연산은 bit 연산인데, 1의 보수 연산을 수행해줍니다.
이 과정에서 모든 소수점을 버리게 됩니다.
`~~`처럼 표현할 경우 1의 보수 연산을 2번 수행합니다. 즉, 소수점만 버리고 원래대로 되돌려놓는거죠 ㅎㅎ

그런데 숫자가 매우 커질 경우에는 제대로 작동하지 않을 때가 있습니다.

JunilHwang commented on src/js/index.js:
```diff
@@ -0,0 +1,104 @@
+  value >= 0 ? Math.floor(value) : Math.ceil(value);
+
+const sum = () => {
+  const [first, second] = valueList(curOperator()).map(Number);
+
+  const lookupTable = {
+    [plus]: first + second,
+    [minus]: first - second,
+    [division]: roundDown(first / second),
+    [multiplication]: roundDown(first * second),
+  };
+
+  return lookupTable[curOperator()];
```
curOperator을 수행할 경우 dom에 접근하는 것 같습니다.
한 번 dom의 값을 가져온다음에 변수에 저장해놓으면 어떨까요?
DOM과 관련된 연산이 JS에서 제일 많은 비용을 차지한답니다!

JunilHwang commented on src/js/index.js:
```diff
@@ -0,0 +1,104 @@
+const sum = () => {
+  const [first, second] = valueList(curOperator()).map(Number);
+
+  const lookupTable = {
+    [plus]: first + second,
+    [minus]: first - second,
+    [division]: roundDown(first / second),
+    [multiplication]: roundDown(first * second),
+  };
```
더 깔끔해졌네요 👍

JunilHwang reviewed (approved):
안녕하세요 영서님!
코로나 때문에 무척 힘드셨군요.. ㅠㅠ 부디 빠르게 회복하시길 기원합니다 🙏

그럼에도 불구하고 피드백 반영 무척 잘 해주셨네요 👍👍👏👏👏
그런데 아직 더 개선할 수 있는 부분들이 많이 있는 상태입니다.
다만 자동차 미션은 온보딩 성격의 미션이기 때문에 빠르게 마무리 하는 것을 목적으로 하고 있습니다 ㅎㅎ
그래서 이번에 남긴 내용들은 다음 미션을 수행하실 때 참고하시면 좋을 것 같습니다.

제일 신경써야 할 부분은 다음과 같습니다.
일단 View(그리는 부분)와 Domain(연산하는 부분)이 뒤섞여 있으면 유지보수하기가 무척 힘듭니다.
한 번에 두 가지를 모두 신경써야 하고,
View와 Domain의 변화율이 다르기 때문입니다.
UI만 변경해야 할 경우가 있고, 도메인 로직만 변경해야 하는 경우가 있는데 이게 뒤섞여 있을 경우 관리하기가 무척 힘들어집니다.
즉, 유지보수하기가 힘들다는 이야기인거죠 ㅎㅎ
이 과정은 클린코드에 대한 과정이고, 클린코드의 목적은 결국 `유지보수` 입니다.
때문에 지속적인 고민을 계속해서 해보세요!

다음 미션인 로또 미션의 경우 도메인이 무척 복잡한 미션입니다.
일단 로또 라는 도메인에 대해서 정확히 이해가 필요하고, 로또 라는 개념을 View에서 어떻게 분리할 수 있을지 고민해보면서 수행해보세요!

짧은 기간이었지만 작게나마 도움이 되었길 바랍니다 🙏
수고하셨습니다!
무사 완주 기원합니다 😁
